### ISSUE TITLE
RFC: Discrete tensors in continuous domains

### ISSUE THREAD

**Issue by willow-ahrens** [willow-ahrens/Finch.jl#230]
Reposting some discussion from https://github.com/willow-ahrens/Finch.jl/pull/217 in an issue here:

Is the following behavior intended?

```julia
julia> x = @fiber(srl(e(0.0)), [0, 0, 1, 1, 1, 0, 0])
    SparseRLE (0.0) [1:7]
    ├─[3:3]: 1.0
    ├─[4:4]: 1.0
    ├─[5:5]: 1.0
```

@nullplay says:

This was intended after some consideration. Because I want to view integer tensors as a pinpoint coordinate ([a,a]) in our continuous indices, an automatic merging (A[3:5] = 1) would semantically result in different tensor in continuous domain.

I think you've raised a very interesting question: How should we interpret the expression A[2:4]=1 in the context of a continuous tensor field? We have previously discussed two options for interpreting existing integer tensors in a field: (1) a pinpoint interval ([n,n]) and (2) the interval with length 1 ([n, n+1)). So far, I believe that interpreting the existing tensor as a pinpoint field is the correct approach.

However, I have a concern regarding the syntax A[2:4]=1. It seems like this syntax would insert a single piece-wise constant value of 1 into the interval [2,4) in the field A. However, from a semantic standpoint, is A[2:4]=1 actually inserting three piece-wise constants into three discontinuous singleton intervals ([2,2], [3,3], [4,4]) in the integer world? If that is the case, should we consider introducing a different syntax to represent the insertion of a single([2,4)) piece-wise constant value?

**Comment by willow-ahrens** [comment]
I think it's okay to have discrete runs: A[2:4] = 7 means that all of the integers between 2 and 4 are 7, and B[2.0:4.0] = 7 means that all the reals between 2.0 and 4.0 are 7. C[i] += A[i] * B[i] is undefined because we never specified the value of A at e.g. 2.5. Thus, we should say either C[i] += coalesce(A[i], 0) * B[i] or C[i] += A[floor(i)] * B[i] to request the appropriate behavior.

Like, I think that a tensor with integer index types should be understood as defined only on integers, same for floats/reals

**Comment by nullplay** [comment]
> I think it's okay to have discrete runs: A[2:4] = 7 means that all of the integers between 2 and 4 are 7, and B[2.0:4.0] = 7 means that all the reals between 2.0 and 4.0 are 7.

> Like, I think that a tensor with integer index types should be understood as defined only on integers, same for floats/reals

I love the suggestion of interpreting an integer index type as representing discrete runs (`A[2:4]` → {[2,2], [3,3], [4,4]}) and a float index type as representing continuous runs on reals (`A[2.0:4.0]` → {[2.0,4.0]}). However, one implementation concern arises when dealing with real numbers that are actually integers, such as 2.0, which might be stored as an `int` type rather than a `float` in order to utilize faster integer arithmetic on CPU.

> C[i] += A[i] * B[i] is undefined because we never specified the value of A at e.g. 2.5. Thus, we should say either C[i] += coalesce(A[i], 0) * B[i] or C[i] += A[floor(i)] * B[i] to request the appropriate behavior.

Regardless of whether we are dealing with discrete or continuous runs, I believe `C[i] += A[i] * B[i]` should work without any modifications like coalesce or floor. Instead of assuming the undefined region in discrete runs, we can treat it as a filled value (typically 0). Therefore, internally, discrete integer runs essentially define a stepper of spikes. As a result, `C[i] += A[i] * B[i]` will only perform effective computations on `i = {2.0, 3.0, 4.0}`.

**Comment by willow-ahrens** [comment]
> However, one implementation concern arises when dealing with real numbers that are actually integers, such as 2.0, which might be stored as an int type rather than a float in order to utilize faster integer arithmetic on CPU.

Perhaps if we want to interpret the integer as a real, we could use a wrapper type to declare it? Julia has several FixedPoint types, for instance (wrappers that allow you to represent reals as integers for efficiency). E.g. https://github.com/JuliaMath/FixedPointNumbers.jl would make this representation more clear, correct, and efficient.

> Regardless of whether we are dealing with discrete or continuous runs, I believe `C[i] += A[i] * B[i]` should work without any modifications like coalesce or floor. 

Okay! Let's say that `C[i] += A[i] * B[i]` interprets undefined values as the default, and let's say that `C[i] += A[floor(i)] * B[i]` gives the behavior of returning the nearest integer value.

**Comment by willow-ahrens** [comment]
@nullplay is this issue solved? What would it take to consider it solved?

**Comment by nullplay** [comment]
Resolved in #219
